src/canisters/frontend/ic-certified-assets/src/state_machine.rs

```diff
@@ -36,12 +44,137 @@ pub struct AssetEncoding {
     pub sha256: [u8; 32],
 }
 
+#[derive(Deserialize, CandidType, Serialize, Debug, Default, Clone, PartialEq, Eq)]
+pub struct RedirectUrl {
+    pub(crate) host: Option<String>,
+    pub(crate) path: Option<String>,
+}
+
+#[derive(Default, Clone, Debug, CandidType, Deserialize)]
+pub struct AssetRedirect {
+    pub from: Option<RedirectUrl>,
+    pub to: RedirectUrl,
+    pub user_agent: Option<Vec<String>>,
+    pub response_code: u16,
+}
+
+impl AssetRedirect {
+    pub fn is_valid(&self) -> Result<(), String> {
+        if self.from.is_some() {
+            if self.from.as_ref().unwrap().host.is_none()
+                && self.from.as_ref().unwrap().path.is_none()
+            {
+                return Err("AssetRedirect.from must have a host or path".to_string());
+            }
+        }
+        if self.to.host.is_none() && self.to.path.is_none() {
+            return Err("AssetRedirect.to must have a host or path".to_string());
+        }
+        Ok(())
+    }
+
+    fn matches_user_agent(&self, user_agent: &str) -> bool {
+        // if user-agent coming in HTTP Request is not found in list of user-agents from redirect config
+        // then we're returning None, otherwise proceed.
+        if self.user_agent.as_ref().map_or(false, |cfg_user_agents| {
+            if cfg_user_agents.is_empty() {
+                return false;
+            }
+            // TODO: performance hit for long lists. Consider using:
+            // - https://docs.rs/regex/latest/regex/struct.RegexSet.html
+            //   > The key advantage of using a regex set is that it will
+            //   > report the matching regexes using a single pass through
+            //   > the text. If one has hundreds or thousands of regexes
+            //   > to match repeatedly (like a URL router for a complex
+            //   > web application or a user agent matcher), then a regex set
+            //   > can realize huge performance gains.
```

Comment:
The intended use case for this is to specify a list of web crawlers, to be able to redirect them efficiently. The list of all known crawlerss isn't exactly short https://www.useragentstring.com/pages/useragentstring.php?typ=Crawler , so I think it may have sense to implement this using `RegexSet`. I haven't checked the performance or anything, but I suspect there is some overhead involved in spawning such object as `RegexSet` if `user_agent` would be a short list. 

I did check if it's possible to somehow initialize such object client side, and (again)somehow transfer raw bytes which would serve as a blueprint for regex-automata, but couldn't find anything like that. It's possible my understanding about how regex engine works is off (so e.g. there is no step that transforms list of regexes into - as I called it - raw bytes blueprint)

I'd love to read your opinion